### ISSUE TITLE
Uploading files larger than 5GB fails

### ISSUE THREAD

**Issue by jakrams** [Azure/azure-sdk-for-cpp#3492]
**Query/Question**

In our application file uploads that are larger than 5GB consistently fail. We are using the BlockBlobClient interface. First we use StageBlock to upload the file in 4 MB pieces. This seems to fully succeed, we never get an exception here and the entire file, no matter how large it is, will be staged.
Finally we use CommitBlockList to finish the file. Anything below 5 GB is working fine, but for larger files we get the error message "couldn't write XML".

I have tried this with various file sizes and so far have narrowed it down to roughly this. Uploading a file of EXACTLY 5GB+128MB (1024*1024*(128+1024*5)) still works. Uploading 5.5 GB doesn't work.

Does this ring a bell somehow? With 4 MB blocks this is about 1300 blocks that will be committed, which doesn't seem to be close to any limit.

**Setup (please complete the following information if applicable):**
 - OS: Windows 10
 - IDE : VS 2022
 - Version of the Library used: latest from vcpkg
 
 **Information Checklist**
 Kindly make sure that you have added all the following information above and checkoff the required fields otherwise we will treat the issuer as an incomplete report
- [x] Query Added
- [x] Setup information Added

**Comment by Jinming-Hu** [comment]
I tested with code below to create a block blob with 1600 4MiB blocks. Blob can be successfully created.


```C++
// Copyright (c) Microsoft Corporation. All rights reserved.
// SPDX-License-Identifier: MIT

#include <cstdio>
#include <iostream>
#include <stdexcept>
#include <thread>

#include <azure/storage/blobs.hpp>
#include <azure/core/base64.hpp>
#include <azure/core/http/curl_transport.hpp>

int main()
{
    using namespace Azure::Storage::Blobs;

    const std::string connectionString = "YOUR_CONNECTION_STRING";
    const std::string containerName = "sample-container";
    const std::string blobName = "sample-blob";
    const int numBlocks = 1600;
    const int concurrency = 8;

    BlobClientOptions options;
    options.Transport.Transport = std::make_shared<Azure::Core::Http::CurlTransport>();

    auto containerClient
        = BlobContainerClient::CreateFromConnectionString(connectionString, containerName, options);
    containerClient.CreateIfNotExists();

    BlockBlobClient blobClient = containerClient.GetBlockBlobClient(blobName);
    blobClient.DeleteIfExists();

    std::vector<uint8_t> buffer(4 * 1024 * 1024, 'a');

    auto getBlockId = [](int id) -> std::string {
        std::string s = std::to_string(id);
        s.resize(10, 'p');  // padding to same length
        return Azure::Core::Convert::Base64Encode(std::vector<uint8_t>(s.begin(), s.end()));
    };

    auto uploadBlock = [&](int i) {
        const std::string blockId = getBlockId(i);
        Azure::Core::IO::MemoryBodyStream memoryBodyStream(buffer.data(), buffer.size());
        blobClient.StageBlock(blockId, memoryBodyStream);
    };

    std::atomic<int> id{ 0 };

    std::vector<std::thread> threadHandles;
    for (int i = 0; i < concurrency; ++i) {
        threadHandles.emplace_back(std::thread([&]() {
            while (true) {
                int i = id.fetch_add(1);
                if (i >= numBlocks) {
                    break;
                }
                try {
                    uploadBlock(i);
                }
                catch (std::exception& e) {
                    std::cout << e.what() << std::endl;
                }
                std::cout << std::to_string(i) + "\n" << std::flush;
            }
            }
        ));
    }
    for (auto& th : threadHandles) {
        th.join();
    }

    std::vector<std::string> blockIds;
    for (int i = 0; i < numBlocks; ++i)
    {
        blockIds.emplace_back(getBlockId(i));
    }
    try {
        blobClient.CommitBlockList(blockIds);
    }
    catch (std::exception& e) {
        std::cout << e.what() << std::endl;
    }

    std::cout << blobClient.GetProperties().Value.BlobSize << std::endl;

    return 0;
}
```

![image](https://user-images.githubusercontent.com/52266153/161234839-3500efb3-6be5-463f-808c-983972a66c63.png)

**Comment by Jinming-Hu** [comment]
@jakrams Can you provide some code that can reproduce the issue?

**Comment by Jinming-Hu** [comment]
Hi @jakrams , I looked through your code. I haven't found what the problem is.

But in the else branch of `#if NEW_STORAGE_SDK`, you just upload the blob from filename. We also have equivalent in new SDK, [UploadFrom](https://github.com/Azure/azure-sdk-for-cpp/blob/b61cd1b91d45011aaf87db52697806f1da7cb5ae/sdk/storage/azure-storage-blobs/inc/azure/storage/blobs/block_blob_client.hpp#L155).

**Comment by jakrams** [comment]
After some more trial and error I have been able to pinpoint this issue more closely:

It seems to be related to the size of the XML file that it writes in CommitBlockList(). This is directly linked to both the used block size and the length of the block ID strings.

I have reduced my block size to 64 KB. At this size, uploading a 64 MB file requires 1024 blocks. Using my GUID string, this still succeeds. However, when I fill up the block ID string to be 60 characters long, it fails with "Failed to write xml." (this is in XmlWriter::Write in the Storage SDK). It succeeds with a 32 MB file.

I haven't attempted to build the SDK myself and debug through it, but the evidence quite clearly points to a fault somewhere in there, failing to create the XML file when it is sufficiently large.
